File: docs/source/examples/python/benchmark.py
from libmuscle import Instance, Message
from ymmsl import Operator

import sys
import time


def driver():
    instance = Instance({
        Operator.O_I: ['out'],
        Operator.S: ['in']})

    # Message size is base_size * scale_base**scale, with scale in
    # the range [0..num_scales).
    base_size = instance.get_setting('base_size', 'int')
    scale_base = instance.get_setting('scale_base', 'int')
    num_scales = instance.get_setting('num_scales', 'int')
    num_repeats = instance.get_setting('num_repeats', 'int')

    # in seconds
    pre_send_delay = instance.get_setting('pre_send_delay', 'float')
    pre_recv_delay = instance.get_setting('pre_recv_delay', 'float')

    while instance.reuse_instance():
        # wait a bit to make sure mirror is running
        time.sleep(0.2)

        for scale in range(num_scales):
            size = base_size * scale_base**scale
            if size > 10 * 1024**3:
                raise RuntimeError('Messages >10GB are not supported')

            message = bytes(size)
            for _ in range(num_repeats):
                time.sleep(pre_send_delay)
                instance.send('out', Message(0.0, None, message))

                time.sleep(pre_recv_delay)
                msg = instance.receive('in')


def mirror():
    instance = Instance({
        Operator.O_I: ['out'],
        Operator.S: ['in']})

    num_scales = instance.get_setting('num_scales', 'int')
    num_repeats = instance.get_setting('num_repeats', 'int')

    # in seconds
    pre_send_delay = instance.get_setting('pre_send_delay', 'float')
    pre_recv_delay = instance.get_setting('pre_recv_delay', 'float')

    while instance.reuse_instance():
        for scale in range(num_scales):
            for _ in range(num_repeats):
                time.sleep(pre_recv_delay)
                msg = instance.receive('in')

                time.sleep(pre_send_delay)
                instance.send('out', msg)


if __name__ == '__main__':
    if len(sys.argv) < 2 or sys.argv[1] not in ('driver', 'mirror'):
        raise RuntimeError(
                'Please specify "driver" or "mirror" as the first argument')
    if sys.argv[1] == 'driver':
        driver()
    else:
        mirror()

File: docs/source/examples/cpp/benchmark.cpp
#include "libmuscle/libmuscle.hpp"
#include "ymmsl/ymmsl.hpp"

#include <cstdint>
#include <iostream>
#include <set>
#include <string>
#include <unistd.h>


using libmuscle::Data;
using libmuscle::Instance;
using libmuscle::Message;
using ymmsl::Operator;

using std::set;
using std::string;


// Sleep for a time period in seconds
void sleep(double time) {
    usleep(static_cast<int64_t>(time * 1e6));
}


// Not optimal, but easily good enough
int64_t powi(int64_t base, int64_t power) {
    int64_t result = 1ll;
    for (int64_t i = 0ll; i < power; ++i)
        result *= base;
    return result;
}


void driver(int argc, char **argv) {
    Instance instance(argc, argv, {
            {Operator::O_I, {"out"}},
            {Operator::S, {"in"}}});

    // Message size is base_size * scale_base**scale, with scale in
    // the range [0..num_scales).
    int64_t base_size = instance.get_setting_as<int64_t>("base_size");
    int64_t scale_base = instance.get_setting_as<int64_t>("scale_base");
    int64_t num_scales = instance.get_setting_as<int64_t>("num_scales");
    int64_t num_repeats = instance.get_setting_as<int64_t>("num_repeats");

    // in seconds
    double pre_send_delay = instance.get_setting_as<double>("pre_send_delay");
    double pre_recv_delay = instance.get_setting_as<double>("pre_recv_delay");

    std::vector<char> buffer(base_size * powi(scale_base, num_scales - 1));

    while (instance.reuse_instance()) {
        // wait a bit to make sure mirror is running
        sleep(0.2);

        for (int64_t scale = 0; scale < num_scales; ++scale) {
            int64_t size = base_size * powi(scale_base, scale);
            if (size > 10ll * powi(1024, 3))
                throw std::runtime_error("Messages > 10GB are not supported");

            for (int64_t i = 0; i < num_repeats; ++i) {
                sleep(pre_send_delay);
                instance.send("out", Message(
                            0.0, Data::byte_array(buffer.data(), size)));
                sleep(pre_recv_delay);
                auto msg = instance.receive("in");
            }
        }
    }
}


void mirror(int argc, char **argv) {
    Instance instance(argc, argv, {
            {Operator::O_I, {"out"}},
            {Operator::S, {"in"}}});

    int64_t num_scales = instance.get_setting_as<int64_t>("num_scales");
    int64_t num_repeats = instance.get_setting_as<int64_t>("num_repeats");

    // in seconds
    double pre_send_delay = instance.get_setting_as<double>("pre_send_delay");
    double pre_recv_delay = instance.get_setting_as<double>("pre_recv_delay");

    while (instance.reuse_instance()) {
        for (int64_t scale = 0; scale < num_scales; ++scale) {
            for (int64_t i = 0; i < num_repeats; ++i) {
                sleep(pre_recv_delay);
                auto msg = instance.receive("in");

                sleep(pre_send_delay);
                instance.send("out", msg);
            }
        }
    }
}


int main(int argc, char * argv[]) {
    if (argc < 2 || !set<string>{"driver", "mirror"}.count(argv[1]))
        throw std::runtime_error(
                "Please specify 'driver' or 'mirror' in the first argument");

    if (argv[1] == string("driver"))
        driver(argc, argv);
    else
        mirror(argc, argv);
}

